File: src/dns.ts
import { domainToASCII } from "node:url";
import { isIP } from "node:net";

export const RECORD_TYPES = ["A", "AAAA", "MX", "TXT", "NS", "CNAME", "SOA", "CAA"] as const;
export type RecordType = typeof RECORD_TYPES[number];
export type Lookup = (name: string, type: RecordType) => Promise<unknown>;
export function parseQuery(query: Record<string, unknown>): { name: string; type: RecordType } {
  if (typeof query.name !== "string" || (query.type !== undefined && typeof query.type !== "string")) throw new Error("name must be a domain; type must be one record type");
  const raw = query.name.trim().replace(/\.$/, "");
  if (/[\s\/\\?#@:％%]/u.test(raw)) throw new Error("name must not contain URL components or escapes");
  const name = domainToASCII(raw).toLowerCase();
  const type = String(query.type ?? "A").toUpperCase() as RecordType;
  if (!name || name.length > 253 || isIP(name) || !name.includes(".") || !name.split(".").every(label => /^[a-z0-9_](?:[a-z0-9_-]{0,61}[a-z0-9_])?$/.test(label))) throw new Error("name must be a valid fully qualified domain name");
  if (!RECORD_TYPES.includes(type)) throw new Error(`type must be ${RECORD_TYPES.join(", ")}`);
  return { name, type };
}
export function createLookup(request: typeof fetch = fetch): Lookup {
  return async (name, type) => {
    const url = new URL("https://dns.google/resolve");
    url.searchParams.set("name", name);
    url.searchParams.set("type", type);
    url.searchParams.set("edns_client_subnet", "0.0.0.0/0");
    // Workers does not implement redirect:"error"; manual still prevents following
    // an upstream redirect, and the non-2xx check below rejects it.
    const response = await request(url, { signal: AbortSignal.timeout(6000), redirect: "manual", headers: { Accept: "application/dns-json" } });
    if (!response.ok) throw new Error("DNS upstream HTTP error");
    const data = await response.json() as { Status?: number; TC?: boolean; Answer?: Array<{ name: string; type: number; TTL: number; data: string }> };
    if (data.Status === 3) throw Object.assign(new Error("No domain"), {code: "ENOTFOUND"});
    if (data.Status !== 0 || data.TC) throw new Error("DNS upstream failed or truncated");
    if (!data.Answer?.length) throw Object.assign(new Error("No records"), {code: "ENODATA"});
    if (!Array.isArray(data.Answer) || !data.Answer.every(r => typeof r.name === "string" && Number.isInteger(r.type) && Number.isInteger(r.TTL) && r.TTL >= 0 && typeof r.data === "string")) throw new Error("Invalid DNS upstream response");
    return data.Answer.map(r => ({ name: r.name, type: r.type, ttl: r.TTL, data: r.data }));
  };
}
export const lookup = createLookup();

File: go-gin/internal/livedns/dns.go
package livedns

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net"
	"net/http"
	"net/url"
	"regexp"
	"strings"
	"time"

	"golang.org/x/net/idna"
)

var recordTypes = map[string]bool{"A": true, "AAAA": true, "MX": true, "TXT": true, "NS": true, "CNAME": true, "SOA": true, "CAA": true}
var labelPattern = regexp.MustCompile(`^[a-z0-9_](?:[a-z0-9_-]{0,61}[a-z0-9_])?$`)

var errNoRecords = errors.New("dns record not found")

type DNSRecord struct {
	Name string `json:"name"`
	Type int    `json:"type"`
	TTL  int    `json:"ttl"`
	Data string `json:"data"`
}

type Resolver func(context.Context, string, string) ([]DNSRecord, error)

func parseQuery(rawName, rawType string) (string, string, error) {
	rawName = strings.TrimSuffix(strings.TrimSpace(rawName), ".")
	if rawName == "" || strings.ContainsAny(rawName, " /\\?#@:%") {
		return "", "", errors.New("name must be a domain")
	}
	name, err := idna.Lookup.ToASCII(rawName)
	if err != nil {
		return "", "", errors.New("name must be a valid domain")
	}
	name = strings.ToLower(name)
	if len(name) > 253 || net.ParseIP(name) != nil || !strings.Contains(name, ".") {
		return "", "", errors.New("name must be a fully qualified domain")
	}
	for _, label := range strings.Split(name, ".") {
		if !labelPattern.MatchString(label) {
			return "", "", errors.New("name must be a valid fully qualified domain")
		}
	}
	typ := strings.ToUpper(rawType)
	if typ == "" {
		typ = "A"
	}
	if !recordTypes[typ] {
		return "", "", errors.New("unsupported DNS record type")
	}
	return name, typ, nil
}

func googleResolver(client *http.Client) Resolver {
	return func(ctx context.Context, name, typ string) ([]DNSRecord, error) {
		u, _ := url.Parse("https://dns.google/resolve")
		q := u.Query()
		q.Set("name", name)
		q.Set("type", typ)
		q.Set("edns_client_subnet", "0.0.0.0/0")
		u.RawQuery = q.Encode()
		requestCtx, cancel := context.WithTimeout(ctx, 6*time.Second)
		defer cancel()
		req, _ := http.NewRequestWithContext(requestCtx, http.MethodGet, u.String(), nil)
		req.Header.Set("Accept", "application/dns-json")
		resp, err := client.Do(req)
		if err != nil {
			return nil, err
		}
		defer resp.Body.Close()
		if resp.StatusCode < 200 || resp.StatusCode >= 300 {
			return nil, fmt.Errorf("DNS upstream HTTP %d", resp.StatusCode)
		}
		var body struct {
			Status int         `json:"Status"`
			TC     bool        `json:"TC"`
			Answer []DNSRecord `json:"Answer"`
		}
		if err := json.NewDecoder(resp.Body).Decode(&body); err != nil {
			return nil, err
		}
		if body.Status == 3 || (body.Status == 0 && len(body.Answer) == 0) {
			return nil, errNoRecords
		}
		if body.Status != 0 || body.TC {
			return nil, errors.New("DNS upstream failed or truncated")
		}
		for _, record := range body.Answer {
			if record.Name == "" || record.Type <= 0 || record.TTL < 0 || record.Data == "" {
				return nil, errors.New("invalid DNS upstream response")
			}
		}
		return body.Answer, nil
	}
}
